client/src/main/java/io/atomix/copycat/client/session/ClientSession.java

```diff
@@ -580,12 +589,30 @@ private ClientSession resetConnection() {
    */
   private ClientSession resetMembers() {
     if (connectMembers.isEmpty() || connectMembers.size() < members.size() - 1) {
-      connectMembers = connectionStrategy.getConnections(leader, new ArrayList<>(members));
+      connectMembers = new ArrayList<>(members);
     }
     return this;
   }
 
   /**
+   * Selects the next member to which to send a request.
+   */
+  private Address selectMember() {
+    // Remove the next member that meets the strategy from the members list.
+    Iterator<Address> iterator = connectMembers.iterator();
+    while (iterator.hasNext()) {
+      Address member = iterator.next();
+      if (connectionStrategy.connect(member, leader)) {
+        iterator.remove();
```

Comment:
Actually, it should do that. Effectively, what happens is the client creates the `connectMembers` list. It first tries to connect to members that are allowed by the strategy, and then attempts to connect to members not allowed by the strategy. This ensures that e.g. the `FOLLOWER` strategy can connect to the leader of the cluster is resized down to one node.

So, the `connectMembers` list is used essentially for one iteration of attempts. Once allbers have been attempted (the members list is empty) it will schedule a retry. This is from `ClientSession` (I'm on my phone right now and having trouble linking :-)

```java
    // If we've run out of servers to which to attempt to connect, determine whether we should expire the
    // session based on the responses from servers with which we did successfully communicate and the
    // time we were last able to successfully communicate with a correct server process. The failureTime
    // indicates the first time we received a NO_LEADER_ERROR from a server.
    if (connectMembers.isEmpty()) {
      // If open checks are not being performed, don't retry connecting to the servers. Simply fail.
      if (!checkOpen) {
        LOGGER.warn("Failed to connect to cluster");
        future.completeExceptionally(new IllegalStateException("session not open"));
      }
      // If retries have already been scheduled, queue a callback to be called to retry the request.
      else if (retryFuture != null) {
        retries.add(() -> {
          LOGGER.debug("Retrying: {}", request);
          request(request, future, true, true);
        });
      }
      // If all servers indicated that no leader could be found and a session timeout has elapsed, time out the session.
      else if (failureTime > 0 && failureTime + timeout < System.currentTimeMillis()) {
        LOGGER.warn("Lost session");
        resetConnection().onExpire();
        future.completeExceptionally(new IllegalStateException("session expired"));
      }
      // If not all servers responded with a NO_LEADER_EXCEPTION or less than a session timeout has expired,
      // schedule a retry to attempt to connect to servers again.
      else {
        LOGGER.warn("Failed to communicate with cluster. Retrying");
        retryFuture = context.schedule(Duration.ofMillis(200), this::retryRequests);
        retries.add(() -> request(request, future, true, true));
      }
      return future;
    }
```
So, the logic in the event that all members have been tried is essentially this:
* If no session is open (`checkOpen` is false) simply fail
* If a session is open and less than a session timeout has elapsed, schedule a retry. This ensures clients don't pound servers during leadership changes, but it also ensures that clients sessions will expire when they can't reach the cluster. This is necessary to ensure that e.g. a client that thinks it has a lock will release it even if it can't reach the cluster (which may give the lock to another client).

But first, back to `connectMembers`. For each retry, `connectMembers` is reset. So, that means clients will repeatedly retry all the servers until they believe their session to be expired. See `resetMembers` which resets the members list to a full members list any time a request is successful.

But I do think there is some work here that needs to be done. Currently, clients only expire their sessions if all servers respond with a `NO_LEADER_ERROR`. But in reality, Copycat will never expire a session when no leader is present, and so clients sessions can persist through arbitrary leadership changes or lengthy periods of loss of quorum. What needs to be determined is when to expire the client's session in cases where it simply can't reach the cluster, the goal being to ensure that two Atomix clients with an open `LeaderElection` resource don't believe themselves to be the leader simultaneously (when one client can't send keep alive requests and so another client gets elected). It may be impossible to guarantee that because we have to rely on system clocks, but expiring clients' sessions when they fail to send keep alive requests is the next best option.